### PR TITLE
Add a filter for unused annotation processors.

### PR DESCRIPTION
**Pull request by stephanenicolas**
Solves #126 . 

There was also a little issue in a different test, I hope you're ok that I fix it here.

### CODE REVIEW

File: src/test/kotlin/com/autonomousapps/internal/ConsoleReportTest.kt [hunk 1]
```diff
@@ -287,10 +289,56 @@ class ConsoleReportTest {
     val consoleReport = ConsoleReport.from(computedAdvice = computedAdvice)
 
     // Then
-    val adviceApi = consoleReport.changeToApiAdvice
-    assertTrue(adviceApi.isEmpty(), "Expected $adviceApi to be empty")
-    val adviceImpl = consoleReport.changeToImplAdvice
-    assertTrue(adviceImpl.isEmpty(), "Expected $adviceImpl to be empty")
+    val adviceCompileOnly = consoleReport.compileOnlyDependencies
```
Comment (stephanenicolas) [inline src/test/kotlin/com/autonomousapps/internal/ConsoleReportTest.kt:292]: Here is the other test that I fixed.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt [hunk 2]
```diff
@@ -87,13 +98,23 @@ data class Advice(
...
    */
   fun isChange() = fromConfiguration != null && toConfiguration != null && !isCompileOnly()
+
+  /**
+   * An advice is "processors-advice" if it is declared on a k/apt
+   * or annotationProcessor configuration.
+   */
+  fun isProcessor() = fromConfiguration != null &&
+    toConfiguration == null &&
+    (fromConfiguration.endsWith("apt", ignoreCase = true)
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/advice/Advice.kt:118]: I appreciate this attempt to be more general (apt vs kapt), but actually we only check for kapt currently. I've made no attempt to be compatible with the old `apt`. And I'm still not convinced we need to differentiate processor-advice from remove-advice like this.

File: src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt [hunk 3]
```diff
@@ -114,4 +119,17 @@ internal class FilterSpec(
...
+  val unusedProcsAdviceFilter: (Dependency) -> Boolean = { dependency ->
+    if (anyBehavior is Ignore || unusedProcsBehavior is Ignore) {
+      // If we're ignoring everything or just ignoring all compileOnly dependencies, then do that
+      false
+    } else if (anyBehavior.filter.plus(unusedProcsBehavior.filter).contains(dependency.identifier)) {
+      // If we're ignoring some specific dependencies, then do that
+      false
+    } else {
+      // If the dependency is universally ignored, do that
+      universalFilter.predicate(dependency)
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt:132]: Latest master has changed the convention somewhat. These two lines are now generally:
```kotlin
// If the dependency is universally ignored, or is a facade we care about, ignore it
universalFilter.predicate(dependency) && facadeFilter.predicate(hasDependency)
```
although it seems highly unlike an annotation processor would trigger the facade filter. :thinking:

File: src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt [hunk 4]
```diff
@@ -37,13 +37,17 @@ abstract class FailOrWarnTask : DefaultTask() {
...
+  abstract val failOnUnusedProcs: Property<Behavior>
+
   @TaskAction
   fun action() {
     val adviceReports = advice.get().asFile.readText().fromJsonSet<BuildHealth>()
 
     val anyUnusedDependencies = adviceReports.any { (_, advices, _) -> advices.any { it.isRemove() } }
     val anyUsedTransitives = adviceReports.any { (_, advices, _) -> advices.any { it.isAdd() } }
     val anyIncorrectConfigurations = adviceReports.any { (_, advices, _) -> advices.any { it.isChange() } }
+    val anyUnusedProcs = adviceReports.any { (_, advices, _) -> advices.any { it.isChange() } }
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt:50]: do you mean to use `it.isChange()` at the end?

Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt:50]: No, it's a merge issue.
Thx

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt [hunk 5]
```diff
@@ -69,6 +69,17 @@ data class Advice(
       dependency = dependency,
       fromConfiguration = dependency.configurationName, toConfiguration = toConfiguration
     )
+
+    fun processor(dependency: Dependency) = Advice(
```
[the remark this answers is not in the document]
Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/advice/Advice.kt:73]: i/ got to review this but I am not sure to see how everything is going to work. Are you saying they are completely distinct sets of advices all the time and there is not risk to have advices being misinterpreted ?

File: src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt (hunk 3 above)
Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt:132]: Let's keep the facade. Who knows ?

File: src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt (hunk 4 above)
Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt:50]: I don't understand here how you wanna do without introducing a `isProcessor`.

File: src/main/kotlin/com/autonomousapps/internal/models.kt [hunk 6]
```diff
@@ -301,10 +301,10 @@ data class Method internal constructor(val types: Set<String>) {
 }
 
 data class AnnotationProcessor(
-  val dependency: Dependency,
+  override val dependency: Dependency,
```
Comment (stephanenicolas) [inline src/main/kotlin/com/autonomousapps/internal/models.kt:304]: @autonomousapps is it the right way to do it ? `AnnotationProcessor` now implements `HasDependency`.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/advice/Advice.kt:73]: Yes, I consider it non-dangerous. Remove is remove. We distinguish dependency from annotation processor based on a set of well-known configuration names. The only other way is to add a new property to `Advice`, and I already feel like that class is bloated enough. Adding a new factory can only be redundant without a new property (which I'm against).

File: src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt (hunk 4 above)
Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/tasks/FailOrWarnTask.kt]: `isProcessor()` seems ok. It's the new factories for creating `Advice` instances I'm opposed to.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/advice/Advice.kt:73]: Looking at your code again, and I strongly believe we should remove this new method and just keep using the original 
```kotlin
@JvmStatic
fun remove(dependency: Dependency) = Advice(
  dependency = dependency,
  fromConfiguration = dependency.configurationName, toConfiguration = null
)
```
they're identical.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt [hunk 7]
```diff
@@ -87,13 +92,23 @@ data class Advice(
   fun isAdd() = fromConfiguration == null && !isCompileOnly()
 
   /**
-   * An advice is "remove-advice" if it is declared and not used, AND is not `compileOnly`.
+   * An advice is "remove-advice" if it is declared and not used, AND is not `compileOnly`,
+   * AND is not `processor`.
    */
-  fun isRemove() = toConfiguration == null && !isCompileOnly()
+  fun isRemove() = toConfiguration == null && !isCompileOnly() &&!isProcessor()
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/advice/Advice.kt:98]: This is fine. I think you're right and we need some easy way to distinguish dependency-related remove advice from processor-related remove advice. Having an `isProcessor()` method makes sense.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt [hunk 8]
```diff
@@ -87,13 +92,23 @@ data class Advice(
...
    * An advice is "change-advice" if it is declared and used (but is on the wrong configuration),
    * AND is not `compileOnly`.
    */
   fun isChange() = fromConfiguration != null && toConfiguration != null && !isCompileOnly()
+
+  /**
+   * An advice is "processors-advice" if it is declared on a k/apt
+   * or annotationProcessor configuration.
+   */
+  fun isProcessor() = fromConfiguration != null &&
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/advice/Advice.kt:110]: I think that `fromConfiguration != null` is redundant, given lines 112-113, where you test that `fromConfiguration` ends with kapt or annotationProcessor. You might need to do this:
```kotlin
fromConfiguration?.endsWith("kapt", ignoreCase = true) == true
```

File: src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt [hunk 9]
```diff
@@ -114,4 +119,18 @@ internal class FilterSpec(
...
+    val dependency = hasDependency.dependency
+    if (anyBehavior is Ignore || unusedProcsBehavior is Ignore) {
+      // If we're ignoring everything or just ignoring all unused procs dependencies, then do that
+      false
+    } else if (anyBehavior.filter.plus(unusedProcsBehavior.filter).contains(dependency.identifier)) {
+      // If we're ignoring some specific dependencies, then do that
+      false
+    } else {
+      // If the dependency is universally ignored, do that
+      universalFilter.predicate(dependency) && facadeFilter.predicate(hasDependency)
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/advice/filter/FilterSpec.kt:133]: I appreciate that you did this (using `facadeFilter`), but I think we should remove that. For now, at least. I can't think of a scenario where an annotation processor would reasonably be considered a "facade", as I term it. Therefore I'd rather _not_ do this and see if anyone complains and provides a good use-case.

File: src/main/kotlin/com/autonomousapps/internal/models.kt (hunk 6 above)
Comment (autonomousapps) [reply src/main/kotlin/com/autonomousapps/internal/models.kt:304]: Yes, assuming you need to use this somewhere that requires `HasDependency`, this is the right way to do it.

I don't even mind if this stays when you remove the usage that required it (facadeFilter) -- assuming you agree with me there.

File: src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt [hunk 10]
```diff
@@ -180,6 +183,7 @@ abstract class AdviceTask : DefaultTask() {
     unusedDependenciesBehavior = failOnUnusedDependencies.get()
     usedTransitivesBehavior = failOnUsedTransitiveDependencies.get()
     incorrectConfigurationsBehavior = failOnIncorrectConfiguration.get()
+    unusedProcsBehavior= failOnUnusedProcs.get()
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt:186]: nit: missing a space before the `=`.

File: src/main/kotlin/com/autonomousapps/advice/Advice.kt (hunk 8 above)
Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/advice/Advice.kt:110]: I didn't like the 2 `?.`, so I used a let. Let me know if it's ok for you.

File: src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt (hunk 10 above)
Comment (stephanenicolas) [reply src/main/kotlin/com/autonomousapps/tasks/AdviceTask.kt:186]: we badly need ktlint..

**Comment by stephanenicolas** [comment]
I am rebasing and addressing comments. I wanna get another build on CI to debug the failing tests.

**Comment by stephanenicolas** [comment]
@autonomousapps when we get a green build, you should update the [wiki](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior)

**Pull request merged**